Split reth-stages into reth-stages-api and reth-stages

AbnerZheng opened the pull request:
close https://github.com/paradigmxyz/reth/issues/7599

AbnerZheng commented:
@onbjerg PTAL whether it is in the right direction, and if so, is there any more common trait should be extract int reth-stage-api?

mattsse commented on bin/reth/src/commands/stage/dump/execution.rs:
```diff
@@ -10,7 +10,8 @@ use reth_node_ethereum::EthEvmConfig;
 use reth_provider::{ChainSpecProvider, ProviderFactory};
 use reth_revm::EvmProcessorFactory;
-use reth_stages::{stages::ExecutionStage, Stage, UnwindInput};
+use reth_stages::stages::ExecutionStage;
+use reth_stages_api::{Stage, UnwindInput};
```
I'd like to reexport api types from stages, this way we don't need to change anything here

AbnerZheng replied:
Done.

onbjerg commented on crates/stages/src/pipeline/set.rs:
```diff
@@ -1,6 +1,6 @@
-use crate::Stage;
 use reth_db::database::Database;
 use reth_primitives::stage::StageId;
+use reth_stages_api::Stage;
```
this should be moved to the api crate

AbnerZheng replied:
How about "test_utils"? Is it good idea to move them into "stages_api"?

mattsse replied:
imo test-utils should stay in reth-stages

AbnerZheng replied:
It makes sense. But the tests of `pipeline` still relay on `reth_stages::util_utils::TestStage`, and while `pipeline` has been moved into `reth_stages_api`, I moved `reth_stages::util_utils::TestStage` into `reth_stages_api::util_utils::TestStage`, is there any problem?

onbjerg reviewed (changes requested):
lgtm overall but yes, please re-export the api types from the reth-stages crate to simplify the dependencies for crates that require types from both

i'd also like to move the pipeline to the api crate if possible, since this should not depend on any stage impls

i.e. reth-stages is only the impl of the various stages + the default stage sets defines in crates/stages/src/sets.rs

mattsse commented:
sorry about the conflicts -.-

AbnerZheng commented:
> sorry about the conflicts -.-

No worries

onbjerg reviewed (commented):
this looks good to me overall now, will give a final review once we're out of draft. thanks for taking this on 😄

onbjerg commented:
Hey @AbnerZheng, thanks for starting this! I went ahead and applied a few fixes to get it over the line since I need this 😄

mattsse reviewed (approved):
lgtm

perhaps we should consider grouping the stages crates in a new folder `sync/{stages-api,stages}`?

AbnerZheng commented:
> Hey @AbnerZheng, thanks for starting this! I went ahead and applied a few fixes to get it over the line since I need this 😄

Wait a moment, let me remove TODOs I added.

AbnerZheng commented:
> > Hey @AbnerZheng, thanks for starting this! I went ahead and applied a few fixes to get it over the line since I need this 😄
> 
> Wait a moment, let me remove TODOs I added.

Will fire a PR to remove the unnecessary TODO I added soon.

onbjerg commented:
Ah, sorry 🙈
